.github/main.go

```diff
@@ -0,0 +1,167 @@
+		return "sdk " + sdk + " " + action
+	}
+	gh = gh.
+		withSimplePipeline(name("test"), command("test")).
+		withSimplePipeline(name("test (dev-engine)"), command("test"),
+			dagger.GhaWithPipelineOpts{
+				DaggerVersion: ".",
+				Runner:        gh.SilverRunner(true),
+			},
+		).
+		withSimplePipeline(name("lint"), command("lint")).
+		withSimplePipeline(name("lint (dev-engine)"), command("lint"),
+			dagger.GhaWithPipelineOpts{
+				DaggerVersion: ".",
+				Runner:        gh.SilverRunner(true),
+			},
+		)
+	// java doesn't have an automated publish step to test
+	if sdk != "java" {
+		testPublishCommand := command("publish --dry-run=true --tag=$GITHUB_REF")
+		gh = gh.
+			withSimplePipeline(name("test-publish"), testPublishCommand).
+			withSimplePipeline(name("test-publish (dev-engine)"), testPublishCommand,
+				dagger.GhaWithPipelineOpts{
+					DaggerVersion: ".",
+					Runner:        gh.SilverRunner(true),
+				},
+			)
+	}
+	return gh
+}
+
+func (gh *Github) withPipeline(name, command string, opts ...dagger.GhaWithPipelineOpts) *Github {
+	gh.CI = gh.CI.WithPipeline(name, command, opts...)
+	return gh
+}
+
+// Assemble a runner name for a pipeline
+func (gh *Github) Runner(
+	generation int,
+	daggerVersion string,
+	cpus int,
+	singleTenant bool,
+	noSpot bool,
+	dind bool,
+) string {
+	runner := fmt.Sprintf(
+		"dagger-g%d-%s-%dc",
+		generation,
+		strings.ReplaceAll(daggerVersion, ".", "-"),
+		cpus)
+	if dind {
+		runner = runner + "-dind"
+	}
+	if singleTenant {
+		runner = runner + "-st"
+	}
+	if noSpot {
+		// "on demand instances" are the opposite of spot instances in AWS jargon
+		runner = runner + "-od"
+	}
+	// Fall back to default runner if repository is not upstream
+	// (this is GHA DSL and will be evaluated by the GHA runner)
+	return fmt.Sprintf(
+		"${{ github.repository == '%s' && '%s' || '%s' }}",
+		upstreamRepository,
+		runner,
+		defaultRunner,
+	)
+}
+
+// Bronze runner: Multi-tenant spot instance, 4 cpu
+func (gh *Github) BronzeRunner(
+	// Enable docker-in-docker
+	// +optional
+	dind bool,
+) string {
+	return gh.Runner(2, daggerVersion, 4, false, false, dind)
+}
+
+// Silver runner: Multi-tenant spot instance, 8 cpu
+func (gh *Github) SilverRunner(
+	// Enable docker-in-docker
+	// +optional
+	dind bool,
+) string {
+	return gh.Runner(2, daggerVersion, 8, false, false, dind)
+}
+
+// Gold runner: Single-tenant on-demand instance, 16 cpu
+func (gh *Github) GoldRunner(
+	// Enable docker-in-docker
+	// +optional
+	dind bool,
+) string {
+	return gh.Runner(2, daggerVersion, 16, true, true, dind)
+}
+
+// Platinum runner: Single-tenant on-demand instance, 32 cpu
+func (gh *Github) PlatinumRunner(
+	// Enable docker-in-docker
+	// +optional
+	dind bool,
+) string {
+	return gh.Runner(2, daggerVersion, 32, true, true, dind)
+}
+
+// Generate Github Actions pipelines to call our Dagger pipelines
+func (gh *Github) Generate() *dagger.Directory {
+	return gh.CI.Config().Directory(".github")
+}
```

Comment:
A little ask - could we also use the `dirdiff` module to implement a `lint` action here as well (and *maybe* have a pre-defined workflow to make sure this in sync?).

I wish there was a way to avoid needing to have these generated files in the repo at all, I might have some fun ideas for that (hackathon project?), but they involve way too much hackery at the github actions level for now.

Reply:
I actually implemented this already 😁 kudos to @kpenfound for helping figure it out. Unfortunately it's not that great, the best gha can do is generate the files and commit/push them for you. Unlike other CIs it doesn't support continuation.

Reply:
Ahhhh, yeah that's a neat approach. That's annoying it's not great, I think we can just stick with manual invocation for now :cry: (definitely not worth blocking on)

Reply:
I think this gap is a possible insertion point for bridging to pocketCI :) But I don't know exactly how. cc @kpenfound @matipan

Reply:
@jedevc I added a `Check()` function that uses `dirdiff`. But I wasn't sure where to integrate it in the main module, so I didn't wire it in. I could use your guidance on that.

Reply:
Hm, I'd add a new github workflow for right now for this - something like `github lint`? I suppose wiring it into the main module makes sense, but we could also just have a single workflow that directly calls the github module, I'm not hugely biased towards either.